src/crypto/tls/tls_ech.go

```diff
@@ -0,0 +1,684 @@
+// Copyright 2020 Cloudflare, Inc. All rights reserved. Use of this source code
+// is governed by a BSD-style license that can be found in the LICENSE file.
+
+package tls
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/cisco/go-hpke"
+	"github.com/cisco/go-tls-syntax" // TODO(cjpatton): Remove this dep
+	"github.com/cjpatton/ech-mint"
+
+	"golang.org/x/crypto/cryptobyte"
+)
+
+const (
+	// Constants for HPKE operations
+	echLabelAcceptConfirm         = "ech accept confirmation"
+	echHpkeInfoInnerDigest        = "tls13 ech inner digest"
+	echHpkeInfoConfigId           = "tls13 ech config id"
+	echHpkeInfoSetup              = "tls13 ech"
+	echHpkeInfoSetupHrr           = "tls13 ech hrr"
+	echHpkeHrrKeyExportechContext = "tls13 ech hrr key"
+	echHpkeHrrKeyId               = "hrr key"
+	echHpkeHrrKeyLen              = 16
+)
+
+// When providing covertext for the ECH extension, this configuration is
+// used to generate the dummy ECH extension.
```

Comment:
The fields should be randomized with realistic values. If you don't want to do that now, let's add a very clear TODO comment. See https://tlswg.org/draft-ietf-tls-esni/draft-ietf-tls-esni.html#name-grease-extensions

Reply:
Which fields? Presumably you mean the ciphertext (enc, payload)? These are randomized.

Reply:
Not just those two, but also the suite and config_id fields. Your proposed implementation constructs a real HPKE context and encrypts some dummy data, but it would be sufficient to stuff random data in it as described in the IETF draft:
> Set the "suite" field to a supported ECHCipherSuite. The selection **SHOULD vary** to exercise all supported configurations, but MAY be held constant for successive connections to the same server in the same session.

(I am not particularly worried about timing attacks on the difference between doing "real" ECH computations versus generating random ClientECH.)

I am not feeling too strongly about randomizing the suite too and avoiding redundant computations, so if you prefer that go for it.

Reply:
Good point about the config_id. It definitely makes sense to randomize this. The favored suite is likely to be X25519/HKDF-SHA256/AES128-GCM, so I think it's fine to stick with that for the purposes of "don't stick out. I modified the code to reflect this change.